kernel/src/dynamic_process_loading.rs

```diff
@@ -0,0 +1,1242 @@
+
+        let mut processes_start_addresses: [usize; MAX_PROCS] = [0; MAX_PROCS];
+        let mut processes_end_addresses: [usize; MAX_PROCS] = [0; MAX_PROCS];
+
+        self.procs.map(|procs| {
+            for (procs_index, value) in procs.iter().enumerate() {
+                match value {
+                    Some(app) => {
+                        processes_start_addresses[procs_index] = app.get_addresses().flash_start;
+                        processes_end_addresses[procs_index] = app.get_addresses().flash_end;
+                    }
+                    None => {
+                        processes_start_addresses[procs_index] = 0;
+                        processes_end_addresses[procs_index] = 0;
+                    }
+                }
+            }
+        });
+
+        // We compute the closest neighbor to our app such that:
+        // 1. If the new app is placed in between two existing processes, we compute the closest located processes
+        // 2. Once we compute these values, we determine if we need to write a pre pad header, or a post pad header, or both
+        // 3. If there are no apps after ours in the process array, we don't do anything
+
+        // postpad requirement
+        if let Some(next_closest_neighbor) = processes_start_addresses
+            .iter()
+            .filter(|&&x| x > new_app_end_address - 1)
+            .min()
+        {
+            next_app_start_addr = *next_closest_neighbor; // we found the next closest app in flash
+            if self.padding_requirement.is_none() && next_app_start_addr != 0 {
+                self.padding_requirement.set(PaddingRequirement::PostPad);
+            }
+        } else {
+            if config::CONFIG.debug_load_processes {
+                debug!("No App Found after the new app so not adding post padding.");
+            }
+        }
+
+        // prepad requirement
+        if let Some(previous_closest_neighbor) = processes_end_addresses
+            .iter()
+            .filter(|&&x| x < new_app_start_address + 1)
+            .max()
+        {
+            previous_app_end_addr = *previous_closest_neighbor; // we found the next closest app in flash
+            if self.padding_requirement.is_none()
+                && new_app_start_address - previous_app_end_addr != 0
+            {
+                if self.padding_requirement.get() == Some(PaddingRequirement::PostPad) {
+                    self.padding_requirement
+                        .set(PaddingRequirement::PreAndPostPad);
+                } else {
+                    self.padding_requirement.set(PaddingRequirement::PrePad);
+                }
+            }
+        } else {
+            if config::CONFIG.debug_load_processes {
+                debug!("No Previous App Found, so not prepadding the new app.");
+            }
+        }
+        self.previous_app_end_addr.set(previous_app_end_addr);
+        self.next_app_start_addr.set(next_app_start_addr);
```

Comment:
This function is called check_*, but it modifies internal state. That is somewhat confusing.

Reply:
renamed it to `compute_padding_requirements_and_neighbors()`